Major Update.

SebKrantz opened the pull request:
Package is 2x faster, more secure, implemented KWW decomposition, major revision of documentation. Passes R CMD check. I sent you an e-mail.

bquast commented:
Hi Seb,

I appreciate all the work on this. I will have a look at this.

One issue is that CRAN no longer allows the Rstudio extension the way it is currently implemented (write to .GlobalEnv). So if I submit a new version to CRAN I have to remove this. Thoughts?

Bastiaan

SebKrantz commented:
Hi Basti,

thanks, I’ll look at it again in the weekend and come up with a suggestion.
Just a question: Would you mind depending on the matrixStats package? I’ve
already increased WWZ speed 2x just by writing more efficient R code.
matrixStats has C-functions that can sum over subsets of a matrix. So we
can sum without first needing to subset the matrix. I am confident that
depending on matrixStats would give another 2x speedup and I’m happy to
implement that during the weekend.

Best regards,

Sebastian


On Tue 16. Mar 2021 at 15:47, Bastiaan Quast ***@***.***>
wrote:

> Hi Seb,
>
> I appreciate all the work on this. I will have a look at this.
>
> One issue is that CRAN no longer allows the Rstudio extension the way it
> is currently implemented (write to .GlobalEnv). So if I submit a new
> version to CRAN I have to remove this. Thoughts?
>
> Bastiaan
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bquast/decompr/pull/14#issuecomment-800229075>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALOT4URZ73ME3DI7B5GPM3LTD5HPDANCNFSM4Y7AUXFQ>
> .
>

SebKrantz commented:
So I now added the dependency on matrixStats, using functions rowSums2 and colSums2 where possible to efficiently sum over subsets of matrices, and I also added a C-function 'rowmult' to efficiently multiply a vector to each row of a matrix. So in terms of infrastructure we are now setup quite good, speedup is about 3x the original. I still think there could be further improvements by just doing less i.e. coding some things in a smarter way, but I don't want to put too much time into that now. 

Regarding accuracy, in my applications I find that the WWZ does not sum exactly to exports (texpdiff) is not 0. When using the function `wwz2kww(wwz(decompr_object), aggregate = TRUE)` and comparing it to `kww(decompr_object)` I find the only difference to be in term 9 of the KWW, which is the foreign double counted term, the sum of terms 15 (ODC) and 16 (MDC) in the WWZ. So perhaps I introduced a mistake, (I don't think so), or something is wrong with the way we calculate one or both of these terms in the WWZ. 

Sebastian

SebKrantz commented:
Hi Bastiaan, 

it would be really nice if you could find time to review my enhancements so that we can start bringing out an update to the package. 

Thanks, 

Sebastian

bquast commented:
Sorry this took so long, but now with rmarkdown needing to become a dependency, I will need to push a new update. So I will anyway need to delete the GUI functionality. I'll work on it now, I'll let you know when it's done

SebKrantz commented:
Thanks! One note, I added a C function C_rowmult to efficiently multiply
each row of a matrix with a vector. I realized late that matrixStats
provides a function t_tx_OP_y (efficient shorthand for t(t(x) OP y) that
can also do that. My benchmarks give that my function is about 4x faster
than the matrix stats one (might also be because matrix stats is extra
precise). In any case, since I'm using rowSums2 and colSums2 from
matrixstats anyway, you need to decide to keep the C function or use the
matrixstats version. Also check again the accuracy of WWZ, with some
datasets I get residual items (but already with old version of the
package).

```
devtools::load_all()
X <- matrix(rnorm(1e6), 1000)
y <- x[,1]

microbenchmark::microbenchmark(.Call(C_rowmult, X, y), t_tx_OP_y(X, y, "*"))
Unit: milliseconds
                  expr       min        lq     mean    median        uq
 max neval cld
.Call(C_rowmult, X, y)  3.945274  4.093651  5.99098  4.348236  4.991948
21.56312   100  a
  t_tx_OP_y(X, y, "*") 14.369169 15.046573 18.32421 15.996633 19.376066
34.37355   100   b
```

On Sat 8. May 2021 at 16:43, Bastiaan Quast ***@***.***>
wrote:

> Sorry this took so long, but now with rmarkdown needing to become a
> dependency, I will need to push a new update. So I will anyway need to
> delete the GUI functionality. I'll work on it now, I'll let you know when
> it's done
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bquast/decompr/pull/14#issuecomment-835369497>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALOT4UTRUUABYFBXVNYNBQ3TMU5WRANCNFSM4Y7AUXFQ>
> .
>

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virenfrei.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

bquast commented:
Thanks for the clear overview of this introduction, I think your approach is best, hopefully CRAN checks agree

bquast commented:
The new src/decompr.so file that is created when I build it locally, do you add the src folder to .gitignore?

bquast commented:
https://github.com/bquast/decompr/commit/ea5b8fb2bb14778ce5a6383fd128961cf9c4085e ok for you?

SebKrantz commented:
Yeah, you can add *.o *.so *.dll to .gitignore. That's fine, yes.

bquast commented:
I'm running into an error with the Leontief example, do you know what is causing this?

> # Create intermediate object (class 'decompr')
> decompr_object <- load_tables_vectors(leather)
> # exports into their value added origins by country-industry
> leontief(decompr_object)
  Source_Country Source_Industry Using_Country      Using_Industry        FVAX
1      Argentina     Agriculture     Argentina         Agriculture 28.52278143
2      Argentina     Agriculture     Argentina Textile_and_Leather  2.79395126
3      Argentina     Agriculture     Argentina Transport_Equipment  0.35606694
4      Argentina     Agriculture        Turkey         Agriculture  1.81066955
5      Argentina     Agriculture        Turkey Textile_and_Leather  3.11738415
Warning message:
In format.data.frame(if (omit) x[seq_len(n0), , drop = FALSE] else x,  :
  corrupt data frame: columns will be truncated or padded with NAs
